### PR TITLE
Canonicalize CARGO_MANIFEST_DIR

### PR DESCRIPTION
**Pull request by golddranks**
# Objective

- Usually cargo sets `CARGO_MANIFEST_DIR`, and it sets it as an absolute path.
- However, the user might want to set it manually when running the built binary directly to point Bevy to the correct asset path.
- In that scenario, it makes sense to accept relative paths also.
- Currently `FileAssetIo` tends to crash with root set as relpath because of the same problem described here https://github.com/bevyengine/bevy/issues/5689.

## Solution

- Canonicalize `CARGO_MANIFEST_DIR` when setting the root.

### CODE REVIEW

**Comment by bjorn3** [comment]
`std::fs::canonicalize` can fail on Windows. To be specific at least one ram disk program doesn't support it.

**Comment by golddranks** [comment]
Thanks for pointing that out! In that case it makes sense to revert to the original path in case of a failure.

**Review by alice-i-cecile — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
Will this also fix #5689?

**Comment by golddranks** [comment]
@alice-i-cecile Unfortunately, it doesn't. However, I'm planning to play around with `canonicalize` and the co. a bit and see if there is any easy solution.

**Comment by golddranks** [comment]
Hmm, I think I failed to understand what @bjorn3 meant; according to https://github.com/rust-lang/rust/issues/59117 , canonicalized path names on Windows may cause problems even in cases the canonicalization itself succeeds. This made me to change the implementation once more: it now performs the canonicalization only if the path is relative. (`AssetServer` doesn't expect relative paths, so it only tries to fix it in case, when it was already broken, so this shouldn't break anything that currently works.)

**Comment by golddranks** [comment]
Once https://github.com/rust-lang/rust/issues/92750 stabilizes, this should use it instead? (If relative paths are still unsupported by `AssetServer` at that point.)

**Review by afonsolage — APPROVED** [review]

File: crates/bevy_asset/src/io/file_asset_io.rs [hunk 1]
```diff
@@ -64,7 +64,14 @@ impl FileAssetIo {
     /// instead. It's set by cargo when running with `cargo run`.
     pub fn get_base_path() -> PathBuf {
         if let Ok(manifest_dir) = env::var("CARGO_MANIFEST_DIR") {
-            PathBuf::from(manifest_dir)
+            // Some Windows software don't support canonicalized path names, so let's avoid them
+            // unless the path is relative, in which case we currently need to make it absolute
+            // (See more: https://github.com/rust-lang/rust/issues/59117 )
+            if Path::new(&manifest_dir).is_relative() {
```
Comment (cart) [inline crates/bevy_asset/src/io/file_asset_io.rs:70]: I don't think overloading `CARGO_MANIFEST_DIR` with the responsibility of "configuring deployed app asset folder root relative paths" is the right path forward, especially if it is consistently an absolute path within the context of cargo. At the very least, we should have a new environment variable for this scenario.

However, I don't think we should be encouraging "user-configurable" asset roots at all by default. Asset path logic should be determined by the app developer, not the app consumer. The defaults will work in 99.99% of cases and forcing standardization here (by default) is valuable from a tooling and ecosystem perspective.

If you want your game to support non-standard user-configurable asset paths, I think the correct solution is to use `AssetServerSettings::asset_folder` (and maybe add a configurable `custom_root_path: Option<PathBuf>` to the settings). You could wire that up to whatever environment variable you choose.

**Comment by cart** [comment]
Closing this as I've got strong opinions (see my comment above). But feel free keep the conversation going. Happy to listen to counterarguments.

**Pull request closed without merging**